lxd/lifecycle/lifecycle_instance.go

```diff
@@ -0,0 +1,67 @@
+package lifecycle
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// InstanceAction represents a lifecycle event action for instances
+type InstanceAction string
+
+// All supported lifecycle events for instances
+const (
+	InstanceCreated   = InstanceAction("created")
+	InstanceStarted   = InstanceAction("started")
+	InstanceStopped   = InstanceAction("stopped")
+	InstanceShutdown  = InstanceAction("shutdown")
+	InstanceRestarted = InstanceAction("restarted")
+	InstancePaused    = InstanceAction("paused")
+	InstanceResumed   = InstanceAction("resumed")
+	InstanceRestored  = InstanceAction("restored")
+	InstanceDeleted   = InstanceAction("deleted")
+	InstanceRenamed   = InstanceAction("renamed")
+	InstanceUpdated   = InstanceAction("updated")
+)
+
+// Types of instances for which a lifecycle event can occur
+type InstanceType string
+
+// An instance can be either an instance, or a snapshot or backup of an instance
+const (
+	Instance = InstanceType("instance")
+	Snapshot = InstanceType("snapshot")
+	Backup   = InstanceType("backup")
+)
+
+// Lifecycle creates the lifecycle event for an action on an instance
+func (a InstanceAction) Lifecycle(name, projectName string, instanceType InstanceType, ctx map[string]interface{}, op *operations.Operation) api.EventLifecycle {
```

Comment:
signature should be `func (a InstanceAction) Lifecycle(inst instance, ctx map[string]interface{}) api.EventLifecycle {`

You can then just define your own copy of the `instance` interface in the package with the functions you actually need, something like:

```
// Internal copy of the instance interface.
type instance interface {
    Project() string
    Name() string
    Type() instancetype.Type
    Operation() *operations.Operation
}
```

Reply:
> ```
> type instance interface {
>     Type() instancetype.Type
> }
> ```

This one got me scratching my head a bit... If we're already using `Type` to represent whether an instance is a VM or Container, what would be the proper label for if an instance is actually an instance or a backup/snapshot? `Kind`? `InstanceType`? 

Good old naming issues :)

Reply:
Yeah I'm really not keen on introducing another `InstanceType` that has different types that `instance/instancetype`, as we already have a few different ways of representing instance types in different parts of the code base (API vs DB for example). Especially seeing as in this case the constants don't represents instance types (as we have documented them, i.e container, vm).

A snapshot is effectively an instance in its own right (with the type inherited from its parent), it just has a property that indicates its a snapshot.